ipc: ipc_service: pbuf: Add option to disable cache management

nordic-krch opened the pull request:
When packet buffer has producer and consumer on the same core then cache management shall not be used as it will eventually lead to data corruption. Add option to disable cache management and use that in the test which is executed on a single core.

hubertmis commented on subsys/ipc/ipc_service/lib/Kconfig.icmsg:
```diff
@@ -78,3 +78,13 @@ config PBUF
+	  Packet buffer is intended to be used for interprocessor communication
+	  using shared memory. In such case cache must be handled. However, if
+	  packet buffer consumer and producer is on the same core then cache
+	  management shall not be enabled because it may lead to data corruption.
```
I don't like this statement. Unnecessary cache management leads to performance overhead and power waste. But it never should result in data corruption. If it does, we have a bug somewhere.

hubertmis commented on subsys/ipc/ipc_service/lib/pbuf.c:
```diff
@@ -11,6 +11,8 @@
 #include <zephyr/ipc/pbuf.h>
 #include <zephyr/sys/byteorder.h>
 
+#define CONFIG_PBUF_CACHE_HANDLING
+
```
Is this macro in the correct place? It looks suspicous.

hubertmis commented on subsys/ipc/ipc_service/lib/pbuf.c:
```diff
@@ -24,6 +26,20 @@ static uint32_t idx_wrap(uint32_t len, uint32_t idx)
 
+static void cache_flush(void *ptr, size_t data)
+{
+	if (IS_ENABLED(CONFIG_PBUF_CACHE_HANDLING)) {
```
This condition should be per pbuf instance, not global. If a CPU has one pbuf used for internal communication and another for IPC, only one of the instances should have cache management enabled.

On the other hand, we can target scenarios in which all pbuf are internal and then this global config makes sense. But it should be guarded by something like `depends on !IPC`.

In the systems including both internal and IPC pbufs we can by default have cache enabled for all instances. Potentially adding per-isntance cache configuration in the future to improve the performance.

hubertmis reviewed (changes requested):
I think we should find a bug leading to data corruption. Enabling cache handling should never result in data corruption. Only in non-functional problems.

hubertmis commented on subsys/ipc/ipc_service/lib/pbuf.c:
```diff
@@ -152,7 +169,7 @@ int pbuf_read(struct pbuf *pb, char *buf, uint16_t len)
 
 	/* Invalidate wr_idx only, local rd_idx is used to increase buffer security. */
-	sys_cache_data_invd_range((void *)(pb->cfg->wr_idx_loc), sizeof(*(pb->cfg->wr_idx_loc)));
+	cache_flush((void *)(pb->cfg->wr_idx_loc), sizeof(*(pb->cfg->wr_idx_loc)));
```
`data_invd_range` is replaced with `cache_flush`. Should be `cache_invalidate`.

hubertmis commented on subsys/ipc/ipc_service/lib/pbuf.c:
```diff
@@ -175,7 +192,7 @@ int pbuf_read(struct pbuf *pb, char *buf, uint16_t len)
 
 	/* Get packet len.*/
-	sys_cache_data_invd_range(&data_loc[rd_idx], PBUF_PACKET_LEN_SZ);
+	cache_flush(&data_loc[rd_idx], PBUF_PACKET_LEN_SZ);
```
```suggestion
	cache_invalidate(&data_loc[rd_idx], PBUF_PACKET_LEN_SZ);
```

hubertmis commented on subsys/ipc/ipc_service/lib/pbuf.c:
```diff
@@ -203,11 +220,11 @@ int pbuf_read(struct pbuf *pb, char *buf, uint16_t len)
 	uint32_t tail = MIN(blen - rd_idx, len);
 
-	sys_cache_data_invd_range(&data_loc[rd_idx], tail);
+	cache_flush(&data_loc[rd_idx], tail);
```
```suggestion
	cache_invalidate(&data_loc[rd_idx], tail);
```

hubertmis commented on subsys/ipc/ipc_service/lib/pbuf.c:
```diff
@@ -203,11 +220,11 @@ int pbuf_read(struct pbuf *pb, char *buf, uint16_t len)
 
 	if (len > tail) {
-		sys_cache_data_invd_range(&data_loc[0], len - tail);
+		cache_flush(&data_loc[0], len - tail);
```
```suggestion
		cache_invalidate(&data_loc[0], len - tail);
```

hubertmis reviewed (changes requested):
While analysing the code looking for potential bugs, I've spotted some more bugs in this patch

hubertmis commented:
After reviewing the code I see the problem is that `pbuf_write` is not designed to be preempted by `pbuf_read` and I guess that's happening in the test.

If the purpose of this PR is to make the test pass, we have a few options:
* disallow preemption in the `pbuf` API description and reimplement the test to prevent preemption (is the problem in the stress test? Maybe mutex is needed there)
* reimplement `pbuf` to allow preemption (seems like unnecessary memory and performance overhead)
* disable cache management for local-local pub instances only (unnecessary performance overhead if all instances are IPC)
* disallow mixing local-local instances and IPC instances in a single product and disable cache management if all instances are local-local one

I understand this PR is aiming for the last option. If that's the case, it should be better documented.

nordic-krch commented:
@hubertmis 
> This condition should be per pbuf instance, not global

`pbuf` is implemented in `ipc_service` subsystem so it only targets interprocessor communication and having consumer and producer on a single core is only a test. `spsc_pbuf` for example is a generic packet buffer thus it has instance level cache management option.

> I think we should find a bug leading to data corruption. 

I don't think that there is necessarily a bug. In normal circumstances current core only reads or only writes to the buffer so it does not expect that core will fetch data to cache (or flush it).

What about making this option `depends on ZTEST`?

Fixed bugs (i was doing search&replace but must have messed something up).

hubertmis commented on tests/subsys/ipc/pbuf/prj.conf:
```diff
@@ -5,3 +5,4 @@ CONFIG_ZTEST_SHUFFLE=y
 CONFIG_IPC_SERVICE=y
 CONFIG_IPC_SERVICE_ICMSG=y
 CONFIG_PBUF=y
+CONFIG_PBUF_CACHE_HANDLING=n
```
Why bother with the new option and changes in the pbuf implementation instead of 
```suggestion
# pbuf is designed to run writer and reader code on two CPUs.
# The write function cannot be preempted by the read function in a CPU,
# or the cache management could result in data corruption.
# In the tests, the reader and the writer are executed on a single CPU.
# Disable data cache to prevent potential data corruption.
CONFIG_DCACHE=n
```

hubertmis replied:
Thanks. With the comment, it's now obvious to the reader why the cache is disabled and why data corruption is expected with the cache enabled.

nordic-krch committed:
tests: ipc: pbuf: Disable data cache management

Packet buffer is designed to run writer and reader code on two CPUs.
The write function cannot be preempted by the read function in a CPU,
or the cache management could result in data corruption. In the tests,
the reader and the writer are executed on a single CPU. Disable data
cache management to prevent potential data corruption.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

carlocaione commented on tests/subsys/ipc/pbuf/prj.conf:
```diff
@@ -5,3 +5,9 @@ CONFIG_ZTEST_SHUFFLE=y
+# or the cache management could result in data corruption.
+# In the tests, the reader and the writer are executed on a single CPU.
+# Disable data cache to prevent potential data corruption.
+CONFIG_DCACHE=n
```
why not `CONFIG_CACHE_MANAGEMENT`?

nordic-krch replied:
changed to `CONFIG_CACHE_MANAGEMENT=n`. @hubertmis @carlocaione can you take another look?
